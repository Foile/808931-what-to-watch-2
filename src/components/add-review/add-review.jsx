import React from "react";
import {string, number, func, bool, shape, arrayOf} from "prop-types";
import withActiveItem from "../../hocs/with-active-item/with-active-item";
import Header from "../header/header";
import history from "../../history";
import apiDispatcher from "../../reducers/api-dispatcher/api-dispatcher";
import {connect} from 'react-redux';
import {increaseBrightness} from "../../helpers/helpers";
import {compose} from "redux";
import withAuth from "../../hocs/with-auth/with-auth";

const AddReview = (props) => {
  const {addComment, films, match, onChangeActiveItem, activeItem, auth} = props;
  const movie = films.find(({id}) => id === Number(match.params.id));
  if (!movie) {
    return history.push(`/`);
  };

  return <section className="movie-card movie-card--full" style={{background: movie.backgroundColor}}>
    <div className="movie-card__header">
      <div className="movie-card__bg">
        <img
          src={movie.backgroundImage}
          alt={`background ${movie.name}`}
        />
      </div>
      <h1 className="visually-hidden">WTW</h1>
      <Header auth={auth} movie={movie}></Header>
      <div className="movie-card__poster movie-card__poster--small">
        <img src={movie.posterImage} alt={`${movie.name} poster`} width="218" height="327" />
      </div>
    </div>
    <div className="add-review">
      <form action="#" className="add-review__form" onSubmit={
        (evt) => {
          evt.preventDefault();
          const rating = activeItem;
          const comment = evt.target.querySelector(`#review-text`).value;
          addComment(match.params.id, rating, comment);
        }
      }>
        <div className="rating">
          <div className="rating__stars">
            { [1, 2, 3, 4, 5].map((rating) => <React.Fragment key = {`rating-movie-input-${rating}`}>
              <input className="rating__input"
                id={`star-${rating}`}
                type="radio"
                name="rating"
                value={`${rating}`}
                onChange={()=>{
                  onChangeActiveItem(rating);
                }}/>
              <label className="rating__label" htmlFor={`star-${rating}`}>Rating {rating}</label>
            </React.Fragment>
            )}
          </div>
        </div>

        <div className="add-review__text" style={{background: increaseBrightness(movie.backgroundColor, 20)}}>
          <textarea className="add-review__textarea" name="review-text" id="review-text" placeholder="Review text"></textarea>
          <div className="add-review__submit">
            <button className="add-review__btn" type="submit" style={{color: increaseBrightness(movie.backgroundColor, -50)}}>Post</button>
          </div>
        </div>
      </form>
    </div>;
  </section>;
};


AddReview.propTypes = {
  activeItem: number,
  auth: shape({}),
  films: arrayOf(shape({
  })),
  movie: shape({
    name: string,
    genre: string,
    director: string,
    released: number,
    posterImage: string,
    backgroundImage: string,
    runTime: number,
    starring: arrayOf(string)
  }),
  submitHandler: func,
  match: shape({
    params: shape({
      id: string
    })
  })
};

AddReview.defaultProps = {
  activeItem: 0,
  films:[]
};

const mapStateToProps = (state, ownProps) => {
  const res = Object.assign({}, ownProps, {
    films: state.data.allFilms,
    movie: state.data.allFilms && state.data.allFilms.find(({id}) => id === Number(ownProps.match.params.id)),
    auth: state.user.auth,
  });
  return res;
};

const mapDispatchToProps = (dispatch) => ({
  addComment: (movieId, rating, text) => dispatch(apiDispatcher.addComment(movieId, rating, text)),
});

export {AddReview};
const enhance = compose(
  withActiveItem,
  withAuth,
  connect(mapStateToProps, mapDispatchToProps));

export default enhance(AddReview)


