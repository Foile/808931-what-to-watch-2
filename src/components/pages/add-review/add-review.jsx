import React from "react";
import {string, number, func, shape, arrayOf} from "prop-types";
import withActiveItem from "../../../hocs/with-active-item/with-active-item";
import Header from "../../header/header";
import apiDispatcher from "../../../reducers/api-dispatcher/api-dispatcher";
import {connect} from 'react-redux';
import {increaseBrightness} from "../../../helpers/helpers";
import {compose} from "redux";
import withAuth from "../../../hocs/with-auth/with-auth";
import Constants from "../../../const";

const AddReview = (props) => {
  const {onAddComment, movie, match, onChangeActiveItem, activeItem, auth} = props;
  const {id, backgroundColor, backgroundImage, name, posterImage} = movie || {};
  return <section className="movie-card movie-card--full" style={{background: backgroundColor}}>
    <div className="movie-card__header">
      <div className="movie-card__bg">
        <img
          src={backgroundImage}
          alt={`background ${name}`}
        />
      </div>
      <h1 className="visually-hidden">WTW</h1>
      <Header auth={auth} movie={movie}></Header>
      <div className="movie-card__poster movie-card__poster--small">
        <img src={posterImage} alt={`${name} poster`} width="218" height="327" />
      </div>
    </div>
    <div className="add-review">
      <form action="#" className="add-review__form" onSubmit={
        (evt) => {
          evt.preventDefault();
          const rating = activeItem;
          const comment = evt.target.querySelector(`#review-text`).value;
          onAddComment(match.params.id, rating, comment);
        }
      }>
        <div className="rating">
          <div className="rating__stars">
            {Constants.STARS.map((rating) => {
              const isChecked = rating === activeItem;
              return <React.Fragment key = {`rating-movie-${id}-input-${rating}`}>
                <input className={`rating__input`}
                  checked={isChecked}
                  id={`star-${rating}`}
                  type="radio"
                  name="rating"
                  value={`${rating}`}
                  onChange={()=>{
                    onChangeActiveItem(rating);
                  }}/>
                <label className="rating__label" htmlFor={`star-${rating}`}>Rating {rating}</label>
              </React.Fragment>
              ;
            }
            )}
          </div>
        </div>

        <div className="add-review__text" style={{background: increaseBrightness(backgroundColor, 20)}}>
          <textarea className="add-review__textarea" name="review-text" id="review-text" placeholder="Review text"></textarea>
          <div className="add-review__submit">
            <button className="add-review__btn" type="submit" style={{color: increaseBrightness(backgroundColor, -80)}}>Post</button>
          </div>
        </div>
      </form>
    </div>
  </section>;
};

AddReview.propTypes = {
  onAddComment: func,
  onChangeActiveItem: func,
  activeItem: number,
  auth: shape({}),
  movie: shape({
    name: string,
    genre: string,
    director: string,
    released: number,
    posterImage: string,
    backgroundImage: string,
    runTime: number,
    starring: arrayOf(string)
  }),
  match: shape({
    params: shape({
      id: string
    })
  })
};

AddReview.defaultProps = {
  activeItem: Constants.DEF_STAR,
  films: []
};

const mapStateToProps = (state, ownProps) => {
  const res = Object.assign({}, ownProps, {
    films: state.data.allFilms,
    movie: state.data.allFilms && state.data.allFilms.find(({id}) => id === Number(ownProps.match.params.id)),
    auth: state.user.auth,
    activeItem: ownProps.activeItem
  });
  return res;
};

const mapDispatchToProps = (dispatch) => ({
  onAddComment: (movieId, rating, text) => dispatch(apiDispatcher.addComment(movieId, rating, text)),
});

export {AddReview};
const enhance = compose(
    withActiveItem,
    withAuth,
    connect(mapStateToProps, mapDispatchToProps));

export default enhance(AddReview);


