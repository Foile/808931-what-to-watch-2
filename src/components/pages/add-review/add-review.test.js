import React from "react";
import {BrowserRouter} from "react-router-dom";
import renderer from "react-test-renderer";
import {AddReview} from "./add-review";

it(`AddReview correctly renders after relaunch`, () => {
  const movie = {id: 1,
    name: `Macbeth`,
    posterImage: ``,
    previewImage: ``,
    previewVideoLink: ``,
    starring: [],
    genre: `Comedy`,
    backgroundColor: `#ffffff`};

  const tree = renderer
  .create(<BrowserRouter>
    <AddReview
      movie={movie}
      match = {{params: {id: `1`}}}
      onAddComment = {jest.fn()}/>
  </BrowserRouter>)
  .toJSON();
  expect(tree).toMatchSnapshot();
});
