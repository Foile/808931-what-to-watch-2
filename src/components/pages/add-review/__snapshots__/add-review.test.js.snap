// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReview correctly renders after relaunch 1`] = `
<section
  className="movie-card movie-card--full"
  style={
    Object {
      "background": "#ffffff",
    }
  }
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="background Macbeth"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header undefined"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/1"
              onClick={[Function]}
            >
              Macbeth
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <a
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="Macbeth poster"
        height="327"
        src=""
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
        >
          <input
            checked={false}
            className="rating__input"
            id="star-1"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="1"
          />
          <label
            className="rating__label"
            htmlFor="star-1"
          >
            Rating 
            1
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-2"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="2"
          />
          <label
            className="rating__label"
            htmlFor="star-2"
          >
            Rating 
            2
          </label>
          <input
            checked={true}
            className="rating__input"
            id="star-3"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="3"
          />
          <label
            className="rating__label"
            htmlFor="star-3"
          >
            Rating 
            3
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-4"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="4"
          />
          <label
            className="rating__label"
            htmlFor="star-4"
          >
            Rating 
            4
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-5"
            name="rating"
            onChange={[Function]}
            type="radio"
            value="5"
          />
          <label
            className="rating__label"
            htmlFor="star-5"
          >
            Rating 
            5
          </label>
        </div>
      </div>
      <div
        className="add-review__text"
        style={
          Object {
            "background": "#ffffff",
          }
        }
      >
        <textarea
          className="add-review__textarea"
          id="review-text"
          name="review-text"
          placeholder="Review text"
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            style={
              Object {
                "color": "#fefefe",
              }
            }
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
